### DISCUSSION TITLE
Question about Swerve Drivebase FRC Web Component + usage with YAGSL

### DISCUSSION THREAD

**Discussion by tfield** [BroncBotz3481/YAGSL-Example#115] [Q&A] [answered]
Hi BronBots3481 - 

First, thank you for creating and developing YAGSL. We are getting to know it.

**Our setup**
We have our first drivebase set up with mk4i/neo/navx/canencoder.  It's 22"x23". Our absolute encoders are set correctly (thanks to the tips in this discussion board plus a handy YAGSL tutorial video), and we are on a fairly recent (~3 weeks old) version of swervelib.  We based our configuration off the "known good "mk4k/neo/navx" config (but set navx inversion to false).

We are looking for clues as to how to tune the drive, and came across the Swerve Drivebase FRC Web Component. I noticed you were able to instantly identify an error by looking at a short video. I would like to achieve this level of familiarity and recognition.  However, I can't seem to find any documentation on the component, so I was wondering if you could provide an orientation by answering some of the following questions:

**The most basic question**
- Is there documentation for this widget anywhere? If not, perhaps your answers here can serve as a starting point.

**Some questions about the actual widget**
- What do the arrow vectors represent?
- What is the meaning of blue vs red ? (blue seems to be actual and red desired, is this correct?)
- What do the wedges inside each drive module represent?
- What is the meaning of the blue vs the red wedges?
- If I drop a gyro web component onto the dashboard next to my swerve drivebase web component, and then power up my drivebase, I believe I should expect the the robot arrow to be pointed straight up and the gyro to indicate zero.  Turning clockwise should increase the degrees on the gyro and rotate the arrow on the drivebase clockwise, accordingly. Is this correct?

**Some questions about interpretation**
- When the robot is propped up on blocks, we can see wheels turning in a reasonably coordinated fashion, but the more they drive, the more blue and red arrows diverge and the the less the wheels align with the intended direction - even without touching the floor. Why would this be?
- Why, when starting motion from stop (i.e. locked configuration) in a particular direction, would the blue and red arrows not align immediately?

Thank you!  As we become more proficient we look forward to becoming more active in the this community.  In the meantime, thank you for your assistance!

Tony

**Comment by thenetworkgrinch** [comment, marked as answer]
Thank you for posting these questions! I dont think I will answer here but instead take all of your questions and create a page in the gitbook I am creating tomorrow (nicer format I think). I will post here with a link directly to the page, but for now the current gitbook is at https://yagsl.gitbook.io/yagsl 

I am working on condensing all of the documentation onto there and adding more as needed (like for FRC Web Components).

However for a brief answer for some of your questions,

Yes red is "desired" state whereas blue is "actual" state. The slices do represent module position and the arrows show velocity relative to the maximum velocity. It is immensely helpful because you can see what the robot is being told vs what it is actually doing which exasperates the differences like poorly tuned PID's, bad inversions for both drive and angle motors, bad conversion factors, etc. All of these become more visible when you see them long enough. Good luck and there will be more to come on this topic!
